### PR TITLE
Add autotimers, readonly repeat timers and get drive space

### PR DESCRIPTION
**Pull request by phunkyfish**
v3.11.0
- Added: Autotimer support
- Added: Generate ReadOnly Timers for Repeating Timer Rules
- Added: GetDriveSpace

### CODE REVIEW

File: src/Timers.cpp [hunk 1]
```diff
@@ -210,6 +225,218 @@ bool Timers::FindTagInTimerTags(std::string tag, std::string tags)
...
+
+std::vector<AutoTimer> Timers::LoadAutoTimers()
+{
+  std::vector<AutoTimer> autoTimers;
+
+  std::string url; 
+  url = StringUtils::Format("%s%s", vuData.getConnectionURL().c_str(), "autotimer"); 
+
+  std::string strXML;
+  strXML = vuData.GetHttpXML(url);
```
Comment (ksooo) [inline src/Timers.cpp]: You could simplify this and other places slightly by declaring and initializing vars in one go:

`std::string strXML= vuData.GetHttpXML(url);`

And if the var is not modified afterwards a `const` in front of var's type is good style (and can help the compiler to optimize code).

File: src/Timers.cpp [hunk 2]
```diff
@@ -210,6 +225,218 @@ bool Timers::FindTagInTimerTags(std::string tag, std::string tags)
...
+
+  TiXmlElement* pNode = hRoot.FirstChildElement("timer").Element();
+
+  if (!pNode)
+  {
+    XBMC->Log(LOG_DEBUG, "Could not find <timer> element");
+    return autoTimers;
+  }
+
+  for (; pNode != NULL; pNode = pNode->NextSiblingElement("timer"))
```
Comment (ksooo) [inline src/Timers.cpp]: As we're doing C++, you should use `nullptr`, not `NULL`.

File: src/Timers.cpp [hunk 3]
```diff
@@ -210,6 +225,218 @@ bool Timers::FindTagInTimerTags(std::string tag, std::string tags)
...
+    int id;
+    std::string from;
+    std::string to;
+    std::string encoding;
+    std::string searchType;
+    std::string searchCase;
+    std::string avoidDuplicateDescription;
+    std::string searchForDuplicateDescription;
+
+    AutoTimer autoTimer;
```
Comment (ksooo) [inline src/Timers.cpp]: Defining those (well at least the non-trivial) vars outside the for loop generates more performant code as the vars will not be initialized on every iteration.

File: src/Timers.cpp [hunk 4]
```diff
@@ -210,6 +225,218 @@ bool Timers::FindTagInTimerTags(std::string tag, std::string tags)
...
+      }
+    } 
+
+    autoTimer.iWeekdays = 0;
+    
+    TiXmlElement* includeNode = pNode->FirstChildElement("include");
+
+    if (includeNode)
+    {
+      for (; includeNode != NULL; includeNode = includeNode->NextSiblingElement("include"))
```
Comment (ksooo) [inline src/Timers.cpp]: `nullptr`

File: src/Timers.cpp [hunk 5]
```diff
@@ -210,6 +225,218 @@ bool Timers::FindTagInTimerTags(std::string tag, std::string tags)
...
+      }
+    }
+
+    if (autoTimer.iWeekdays != 0)
+    {
+      std::time_t t = std::time(nullptr);
+      std::tm timeinfo = *std::localtime(&t);
+      timeinfo.tm_sec = 0;
+      autoTimer.startTime = 0;
+      if (from != "")
```
Comment (ksooo) [inline src/Timers.cpp]: `if (!from.empty())` is more performant.

File: src/Timers.cpp [hunk 6]
```diff
@@ -210,6 +225,218 @@ bool Timers::FindTagInTimerTags(std::string tag, std::string tags)
...
+      {
+        ParseTime(from, timeinfo);
+        autoTimer.startTime = std::mktime(&timeinfo);
+      }
+
+      timeinfo = *std::localtime(&t);
+      timeinfo.tm_sec = 0;
+      autoTimer.endTime = 0;
+      if (to != "")
```
Comment (ksooo) [inline src/Timers.cpp]: `!to.empty()`

File: src/Timers.cpp [hunk 7]
```diff
@@ -210,6 +225,218 @@ bool Timers::FindTagInTimerTags(std::string tag, std::string tags)
...
+      {
+        ParseTime(to, timeinfo);
+        autoTimer.endTime = std::mktime(&timeinfo);
+      }
+    }
+    else
+    {
+      for (int i = 0; i < 7; i++)
```
Comment (ksooo) [inline src/Timers.cpp]: `7`? Where does this magic number come from. You should define a `static const int MEANINGFUL_NAME = 7;`either here or if that magic 7 is used elsewhere in the header of `Timers`class.

File: src/Timers.cpp [hunk 8]
```diff
@@ -300,14 +540,70 @@ void Timers::GetTimerTypes(std::vector<PVR_TIMER_TYPE> &types)
       PVR_TIMER_TYPE_SUPPORTS_WEEKDAYS         |
       PVR_TIMER_TYPE_SUPPORTS_START_END_MARGIN,
       ""); /* Let Kodi generate the description */
+  types.push_back(*t);
```
Comment (ksooo) [inline src/Timers.cpp]: Did you know that `emplace_back` is to prefer over `push_back` as the former does not make unneded copies of the elements?

File: src/Timers.cpp [hunk 9]
```diff
@@ -300,14 +540,70 @@ void Timers::GetTimerTypes(std::vector<PVR_TIMER_TYPE> &types)
...
+    /* PVR_Timer.iPreventDuplicateEpisodes values and presentation.*/
+    static std::vector< std::pair<int, std::string> > deDupValues =
+    {
+      { AutoTimer::DeDup::DISABLED,                   LocalizedString(30430) },
+      { AutoTimer::DeDup::CHECK_TITLE,                LocalizedString(30431) },
+      { AutoTimer::DeDup::CHECK_TITLE_AND_SHORT_DESC, LocalizedString(30432) },
+      { AutoTimer::DeDup::CHECK_TITLE_AND_ALL_DESCS,  LocalizedString(30433) },
+    };
+
+     /* epg auto search */
```
Comment (ksooo) [inline src/Timers.cpp]: indentation.

File: src/Timers.cpp [hunk 10]
```diff
@@ -335,6 +631,47 @@ void Timers::GetTimers(std::vector<PVR_TIMER> &timers)
...
     tag.iClientIndex = timer.iClientIndex;
+    tag.iParentClientIndex = timer.iParentClientIndex;
+
+    timers.emplace_back(tag);
+  }
+}
+
+void Timers::GetAutoTimers(std::vector<PVR_TIMER> &timers)
+{
+  for (unsigned int i=0; i<m_autotimers.size(); i++)
```
Comment (ksooo) [inline src/Timers.cpp]: `for (const auto& autoTimer : m_autotimers)`

File: src/Timers.cpp [hunk 11]
```diff
@@ -410,8 +859,134 @@ PVR_ERROR Timers::UpdateTimer(const PVR_TIMER &timer)
...
+  unsigned int i=0;
+  while (i<m_autotimers.size())
+  {
+    if (m_autotimers.at(i).iClientIndex == timer.iClientIndex)
+      break;
+    else
+      i++;
+  }
+
+  AutoTimer timerToUpdate = m_autotimers.at(i);
```
Comment (ksooo) [inline src/Timers.cpp]: What happens if you don't find the timer? You will return the last elem in your timers vector. Is this intened?

I would rewrite this like:

```C++
const auto it = std::find_if(
      m_autotimers.cbegin(),
      m_autotimers.cend(),
      [timer](const AutoTimer& autoTimer)
    {
      return autoTimer.iClientIndex  == timer.iClientIndex;
    });

    if (it != m_autotimers.cend())
    {
```

File: src/Timers.cpp [hunk 12]
```diff
@@ -410,8 +859,134 @@ PVR_ERROR Timers::UpdateTimer(const PVR_TIMER &timer)
...
+      }
+
+      if (everyday)
+        includeParams = "&dayofweek=";
+  }
+  else
+  {
+    includeParams = "&dayofweek=";
+  }
+  strTmp += includeParams;
```
Comment (ksooo) [inline src/Timers.cpp]: Haven't I seen (large parts of) this code block before? You should consider refactoring to get rid of copied code.

Comment (ksooo) [reply src/Timers.cpp]: `Timers::AddAutoTimer` looks very similar if not (not checked) identical.

File: src/Timers.cpp [hunk 13]
```diff
@@ -433,12 +1008,53 @@ PVR_ERROR Timers::DeleteTimer(const PVR_TIMER &timer)
...
+  unsigned int i=0;
+  while (i<m_autotimers.size())
+  {
+    if (m_autotimers.at(i).iClientIndex == timer.iClientIndex)
+      break;
+    else
+      i++;
+  }
+
+  AutoTimer timerToDelete = m_autotimers.at(i);
```
Comment (ksooo) [inline src/Timers.cpp]: See my otehr comment. I would rewrite this using `std:.find_if`

File: src/Timers.cpp [hunk 14]
```diff
@@ -433,12 +1008,53 @@ PVR_ERROR Timers::DeleteTimer(const PVR_TIMER &timer)
...
+
+  std::string strTmp;
+  strTmp = StringUtils::Format("autotimer/remove?id=%u", timerToDelete.backendId);
+
+  std::string strResult;
+  if(!vuData.SendSimpleCommand(strTmp, strResult))
```
Comment (ksooo) [inline src/Timers.cpp]: blank between if and ( mising (maybe at other places as well).

File: src/Timers.cpp [hunk 15]
```diff
@@ -486,10 +1104,10 @@ void Timers::TimerUpdates()
...
     if (m_timers.at(i).iUpdateState == VU_UPDATE_STATE_NONE)
     {
-      XBMC->Log(LOG_INFO, "%s Removed timer: '%s', ClientIndex: '%d'", __FUNCTION__, m_timers.at(i).strTitle.c_str(), m_timers.at(i).iClientIndex);
-      m_timers.erase(m_timers.begin()+i);
-      i=0;
-      iRemoved++;
+        XBMC->Log(LOG_INFO, "%s Removed timer: '%s', ClientIndex: '%d'", __FUNCTION__, m_timers.at(i).strTitle.c_str(), m_timers.at(i).iClientIndex);
+        m_timers.erase(m_timers.begin()+i);
+        i=0;
+        iRemoved++;
```
Comment (ksooo) [inline src/Timers.cpp]: You should rewrite all (!) those index based for loops to range based for loops. Then, you get rid of those ugly and inperformant  `.at(i)` stuff.

Example: `for (unsigned int i=0; i<m_timers.size(); i++)` => `for (`const auto& timer : m_timers)`

File: src/Timers.cpp [hunk 16]
```diff
@@ -504,15 +1122,119 @@ void Timers::TimerUpdates()
...
+  }  
+
+  XBMC->Log(LOG_INFO , "%s No of timers: removed [%d], untouched [%d], updated '%d', new '%d'", __FUNCTION__, iRemoved, iUnchanged, iUpdated, iNew); 
+
+  return (iRemoved != 0 || iUpdated != 0 || iNew != 0);
+}
+
+bool Timers::TimerUpdatesAuto()
+{
+  std::vector<AutoTimer> newautotimers = LoadAutoTimers();
```
Comment (ksooo) [inline src/Timers.cpp]: `const std::vector` ...

File: src/Timers.cpp [hunk 17]
```diff
@@ -504,15 +1122,119 @@ void Timers::TimerUpdates()
...
+
+  for (unsigned int i=0; i<m_autotimers.size(); i++)
+  {
+    m_autotimers[i].iUpdateState = VU_UPDATE_STATE_NONE;
+  }
```
Comment (ksooo) [inline src/Timers.cpp]: How feels

```C++
for (auto& autotimer : m_autotimers)
  autoTimer.iUpdateState = VU_UPDATE_STATE_NONE;
```
?

File: src/Timers.cpp [hunk 18]
```diff
@@ -504,15 +1122,119 @@ void Timers::TimerUpdates()
...
+
+          m_autotimers[i].searchPhrase = newautotimers[j].searchPhrase;
+          m_autotimers[i].encoding = newautotimers[j].encoding;
+          m_autotimers[i].searchCase = newautotimers[j].searchCase;
+          m_autotimers[i].searchType = newautotimers[j].searchType;
+          m_autotimers[i].searchFulltext = newautotimers[j].searchFulltext;
+          m_autotimers[i].startAnyTime = newautotimers[j].startAnyTime;
+          m_autotimers[i].endAnyTime = newautotimers[j].endAnyTime;
+          m_autotimers[i].anyChannel = newautotimers[j].anyChannel;
+          m_autotimers[i].deDup = newautotimers[j].deDup;
```
Comment (ksooo) [inline src/Timers.cpp]: All those indexes make me sick. ;-)

File: src/Timers.h [hunk 19]
```diff
@@ -9,6 +9,30 @@
...
+#define AUTOTIMER_AVOID_DUPLICATE_SAME_SERVICE "1"                  //Require Description to be unique - On same service
+#define AUTOTIMER_AVOID_DUPLICATE_ANY_SERVICE "2"                   //Require Description to be unique - On any service
+#define AUTOTIMER_AVOID_DUPLICATE_ANY_SERVICE_OR_RECORDING "3"      //Require Description to be unique - On any service/recording
+
+#define AUTOTIMER_CHECK_SEARCH_FOR_DUP_IN_TITLE "0"                 //Check for uniqueness in - Title
+#define AUTOTIMER_CHECK_SEARCH_FOR_DUP_IN_TITLE_AND_SHORT_DESC "1"  //Check for uniqueness in - Title and Short description
+#define AUTOTIMER_CHECK_SEARCH_FOR_DUP_IN_TITLE_AND_ALL_DESCS ""    //Check for uniqueness in - Title and all descpritions
+
+#define AUTOTIMER_DEFAULT ""
+#define AUTOTIMER_DEFAULT_OMIT(a) (a == AUTOTIMER_DEFAULT)
```
Comment (ksooo) [inline src/Timers.h]: You should prefer static constants over `#defines`, because of type safety, C++ vs C, etc.

File: src/Timers.h [hunk 20]
```diff
@@ -87,6 +113,63 @@ struct Timer
...
+    bChanged = bChanged && (startAnyTime == right.startAnyTime); 
+    bChanged = bChanged && (endAnyTime == right.endAnyTime); 
+    bChanged = bChanged && (anyChannel == right.anyChannel); 
+    bChanged = bChanged && (deDup == right.deDup); 
+    bChanged = bChanged && (!strTitle.compare(right.strTitle));
+
+    return bChanged;
+  }
+
+public:
```
Comment (ksooo) [inline src/Timers.h]: As this is a struct, everything is public - no need for the keywords.

File: src/Timers.h [hunk 21]
```diff
@@ -105,6 +189,14 @@ class Timers
   // functions
   std::vector<Timer> LoadTimers();
   static bool FindTagInTimerTags(std::string tag, std::string tags);
+  static std::string ConvertToAutoTimerTag(std::string tag);
+  std::vector<AutoTimer> LoadAutoTimers();
+  bool CanAutoTimers() const;
+  bool IsAutoTimer(const PVR_TIMER &timer);
```
Comment (ksooo) [inline src/Timers.h]: You should always check whether methods can be made const (not checked for this particular case, though).

File: src/VuData.cpp [hunk 22]
```diff
@@ -235,6 +237,11 @@ const char * Vu::GetServerName()
   return m_strServerName.c_str();  
 }
 
+unsigned int Vu::GetWebIfVersion()
```
Comment (ksooo) [inline src/VuData.cpp]: I bet this metod could be made `const`.

File: src/VuData.cpp [hunk 23]
```diff
@@ -235,6 +237,11 @@ const char * Vu::GetServerName()
...
+{
+  return m_iWebIfVersion;
+}
+
 bool Vu::IsConnected()
```
Comment (ksooo) [inline src/VuData.cpp]: const ?

File: src/VuData.h [hunk 24]
```diff
@@ -27,7 +27,8 @@
 #include "p8-platform/threads/threads.h"
 #include "tinyxml.h"
 
-#define CHANNELDATAVERSION  2
+#define CHANNELDATAVERSION 2
+#define WEBIF_VERSION_NUM(a, b, c) (a << 16 | b << 8 | c)
```
Comment (ksooo) [inline src/VuData.h]: Yikes! A macro! I would prefer a small inline function.

File: src/VuData.h [hunk 25]
```diff
@@ -153,13 +156,14 @@ class Vu  : public P8PLATFORM::CThread
...
   const char * GetServerName();
+  unsigned int GetWebIfVersion();
   bool IsConnected(); 
   std::string getConnectionURL();
   std::vector<std::string> GetLocations();
   std::string GetHttpXML(std::string& url);
   bool SendSimpleCommand(const std::string& strCommandURL, std::string& strResult, bool bIgnoreResult = false);
   std::string URLEncodeInline(const std::string& sSrc);
-
+
```
Comment (ksooo) [inline src/VuData.h]: a blank strolled in here...

**Comment by ksooo** [comment]
Changing code while I'm reviewing is meh! I get lost of track. Hold yer horses.

File: src/Timers.cpp [hunk 26]
```diff
@@ -31,6 +31,43 @@ bool Timer::isRunning(std::time_t *now, std::string *channelName) const
...
+  timeinfo = *std::localtime(&time);
+  std::string parentStartTime = StringUtils::Format("%02d:%02d", timeinfo.tm_hour, timeinfo.tm_min);
+
+  time = parent.endTime;
+  timeinfo = *std::localtime(&time);
+  std::string parentEndTime = StringUtils::Format("%02d:%02d", timeinfo.tm_hour, timeinfo.tm_min);
+
+  bool isChild = true;
+
+  isChild = isChild && (! strTitle.compare(parent.strTitle));
```
Comment (ksooo) [inline src/Timers.cpp]: blank between ! and strTitel

File: src/Timers.cpp [hunk 27]
```diff
@@ -1126,6 +1247,19 @@ bool Timers::TimerUpdatesRegular()
...
   }  
 
+  for (unsigned int i=0; i<m_timers.size();i++)
+  {
+    for (unsigned int j=0; j<m_timers.size();j++)
+    {
+      if ((m_timers.at(i).type == Timer::MANUAL_REPEATING || m_timers.at(i).type == Timer::EPG_REPEATING) &&
+          m_timers.at(j).type == Timer::READONLY_REPEATING_ONCE && m_timers.at(j).isChildOfParent(m_timers.at(i)))
+      {
+        m_timers.at(j).iParentClientIndex = m_timers.at(i).iClientIndex;
```
Comment (ksooo) [inline src/Timers.cpp]: range based for loop vs. index

File: src/Timers.h [hunk 28]
```diff
@@ -84,6 +85,7 @@ struct Timer
 
   bool isScheduled() const;
   bool isRunning(std::time_t *now, std::string *channelName = nullptr) const;
+  bool isChildOfParent(Timer &parent);
```
Comment (ksooo) [inline src/Timers.h]: const or even static?

**Comment by phunkyfish** [comment]
Keep it coming! It's a great opportunity for me to learn. 

And apologies for changing while you were reviewing, I noticed I made a mistake in the versioning of the addon so changed it straight away!

It will take me a while to incorporate the changes so I'll post back here once complete.

**Comment by ksooo** [comment]
Take your time. You're doing an awsome job.

**Comment by phunkyfish** [comment]
Will I bundle all these changes up into a refactoring commit as they are not changing functionality or do you want them fixed up into the existing commits?

The former is easier for me ;)

**Comment by phunkyfish** [comment]
I'm also trying to rewrite the old style for loop used to remove items from an iterator. Is the following best practice or is there a better way?

```
  for (auto existingTimer = m_timers.begin(); existingTimer != m_timers.end(); )
  {
    if (existingTimer->iUpdateState == VU_UPDATE_STATE_NONE)
    {
      existingTimer = m_timers.erase(existingTimer);
      iRemoved++;
    }
    else
    {
      ++existingTimer;
    }
  }   
```

**Comment by ksooo** [comment]
https://en.cppreference.com/w/cpp/algorithm/remove

**Comment by phunkyfish** [comment]
Perfect, thanks!

**Comment by phunkyfish** [comment]
Ok, should be ready to go. I went mad on your suggestions ;)

Thanks for all the great guidance.

File: src/Timers.cpp [hunk 29]
```diff
@@ -758,11 +751,11 @@ void Timers::GetTimers(std::vector<PVR_TIMER> &timers)
...
 }
 
-void Timers::GetAutoTimers(std::vector<PVR_TIMER> &timers)
+void Timers::GetAutoTimers(std::vector<PVR_TIMER> &timers) const
 {
-  for (unsigned int i=0; i<m_autotimers.size(); i++)
+  for (const auto& autoTimer : m_autotimers)
   {
-    AutoTimer &autoTimer = m_autotimers.at(i);
+    //AutoTimer &autoTimer = m_autotimers.at(i);
```
Comment (ksooo) [inline src/Timers.cpp]: Remove?

File: src/Timers.cpp [hunk 30]
```diff
@@ -951,125 +919,141 @@ PVR_ERROR Timers::UpdateTimer(const PVR_TIMER &timer)
...
+    return autoTimer.iClientIndex == timer.iClientIndex;
+  });
+  
+  if (it != m_autotimers.cend())
+  {
+    AutoTimer timerToUpdate = *it;
 
-  AutoTimer timerToUpdate = m_autotimers.at(i);
+    std::string strTmp;
+    strTmp = StringUtils::Format("autotimer/edit?id=%d", timerToUpdate.backendId);
```
Comment (ksooo) [inline src/Timers.cpp]: decl and init of strTmp in one step?

File: src/Timers.cpp [hunk 31]
```diff
@@ -1179,82 +1151,82 @@ bool Timers::TimerUpdatesRegular()
...
+      XBMC->Log(LOG_INFO, "%s New timer: '%s', ClientIndex: '%d'", __FUNCTION__, newTimer.strTitle.c_str(), m_iClientIndexCounter);
+      m_timers.emplace_back(newTimer);
       m_iClientIndexCounter++;
       iNew++;  
     } 
   }  
 
-  for (unsigned int i=0; i<m_timers.size();i++)
+  //Link any Readonly timers to Repeating Timers
+  for (auto& repeatingTimer : m_timers)
```
Comment (ksooo) [inline src/Timers.cpp]: this imo could be `const auto&`

File: src/Timers.cpp [hunk 32]
```diff
@@ -1269,94 +1241,94 @@ bool Timers::TimerUpdatesAuto()
...
+      XBMC->Log(LOG_INFO, "%s New auto timer: '%s', ClientIndex: '%d'", __FUNCTION__, newAutoTimer.strTitle.c_str(), m_iClientIndexCounter);
+      m_autotimers.emplace_back(newAutoTimer);
       m_iClientIndexCounter++;
       iNew++;
     } 
   }
 
-  for (unsigned int i=0; i<m_timers.size();i++)
+  //Link Any child timers to autotimers
+  for (auto& timer : m_timers)
```
Comment (ksooo) [inline src/Timers.cpp]: this could be `const auto&`, imo

File: src/Timers.h [hunk 33]
```diff
@@ -189,17 +187,17 @@ class Timers
...
+  std::vector<AutoTimer> LoadAutoTimers() const;
   bool CanAutoTimers() const;
-  bool IsAutoTimer(const PVR_TIMER &timer);
+  bool IsAutoTimer(const PVR_TIMER &timer) const;
   bool TimerUpdatesRegular();
   bool TimerUpdatesAuto();
-  void ParseTime(const std::string &time, std::tm &timeinfo);
-  bool SendClearSearchForDuplicateValues(int backendId);
+  void ParseTime(const std::string &time, std::tm &timeinfo) const;
+  static std::string BuildAddUpdateAutoTimerIncludeParams(const int weekdays);
```
Comment (ksooo) [inline src/Timers.h]: Making a by-value parameter of primitive type const has no benefits.

File: src/VuData.cpp [hunk 34]
```diff
@@ -1014,22 +997,22 @@ PVR_ERROR Vu::GetChannelGroupMembers(ADDON_HANDLE handle, const PVR_CHANNEL_GROU
...
-int Vu::GetChannelNumber(std::string strServiceReference)  
+int Vu::GetChannelNumber(std::string strServiceReference) const
 {
-  for (unsigned int i = 0;i<m_channels.size();  i++) 
+  for (const auto& channel : m_channels)
   {
-    if (!strServiceReference.compare(m_channels[i].strServiceReference))
-      return i+1;
+    if (!strServiceReference.compare(channel.strServiceReference))
+      return channel.iChannelNumber;
```
Comment (ksooo) [inline src/VuData.cpp]: does this change behavior?

File: src/Timers.cpp (hunk 29 above)
Comment (phunkyfish) [reply src/Timers.cpp]: Done

File: src/Timers.cpp (hunk 30 above)
Comment (phunkyfish) [reply src/Timers.cpp]: Done

File: src/Timers.cpp (hunk 31 above)
Comment (phunkyfish) [reply src/Timers.cpp]: It can be once the argument to isChildOfParent is const which it can be.

File: src/VuData.cpp (hunk 34 above)
Comment (phunkyfish) [reply src/VuData.cpp]: Nope, iChannelNumber is only set here: newChannel.iChannelNumber = m_channels.size()+1; before it's emplaced on the vector.

We never delete channels so ChannelNumber is always equal to vector index+1;

**Comment by phunkyfish** [comment]
Ok, all done ;)

File: src/VuData.cpp (hunk 34 above)
Comment (ksooo) [reply src/VuData.cpp]: okay

**Pull request merged**